onnxruntime/core/providers/cpu/controlflow/loop.cc

```diff
@@ -399,10 +399,33 @@ Status LoopImpl::Execute(FeedsFetchesManager* ffm, const FeedsFetchesManager* ca
       copy_tensor_from_mlvalue_to_output(feeds[i + 2], i);  // skip iter# and cond
     }
 
-    // create empty outputs for loop outputs
-    TensorShape empty;
+    // create empty outputs for loop outputs using the subgraph output shapes for the rank
+    auto& graph_outputs = subgraph_.GetOutputs();
+
     for (int i = num_loop_carried_vars_; i < num_outputs_; ++i) {
-      ORT_IGNORE_RETURN_VALUE(context_.Output(i, empty));
+      std::vector<int64_t> output_dims;
+      output_dims.push_back(0);  // num iterations is first dim
+
+      // get shape from subgraph output if possible to attempt to have the correct rank
+      auto* graph_output = graph_outputs.at(i + 1);  // + 1 as first subgraph output is condition value
+      auto* graph_output_shape = graph_output->Shape();
+
+      if (graph_output_shape) {
+        output_dims.reserve(graph_output_shape->dim_size() + 1);
+
+        // copy the shape. if there are any symbolic dims set them to zero given there's no output data
```

Comment:
I feel that the symbolic dim case is in the same category as the else branch below ... using zero is a heuristic which may or may not match what a user expects in a specific model. May be a similar warning message would be appropriate as in the else-branch.

Reply:
I can add the warning. 

Is it better to leave them as -1?

Reply:
I guess that is also an option. Hard to say which is better. As we discussed elsewhere, there may be an assumption in other places that dim-value will be >= 0, which might break if we leave it as -1. Perhaps converting to 0 is the lesser of evils.

Reply:
One argument for leaving them is that if an output was used as the 'shape' input to a Reshape operator having a -1 is valid. There may be other cases were setting them to 0 is more correct. I will leave them as -1 for now and will refine when needed.

Reply:
Ok. However, from what I remember, a -1 used in the "desired output shape" of a Reshape op indicates a value computable from the output-size (and at most one of the dimensions is allowed to be -1). It is not clear how this output shape will be used in such a context. Other plausible use-cases might be where, e.g., where we "concat" variable-length output produced by a loop with something else (e.g., may be the output of another loop). In this scenario, concat might potentially fail because of dimension mismatch (unless it is smart enough to skip such a check for zero-size tensor, which doesn't sound like the right fix anyway). However, this failure might happen no matter what choice we make above.